### PR TITLE
Drop and merge offline session caches

### LINKED ISSUE #30932: Drop and merge offline session caches
### Description

Drop the offline session and offline client session caches. 

This data can be stored in the online sessions cache using composite keys.

### Discussion

https://github.com/keycloak/keycloak/discussions/30532

### Acceptance criteria

* In scope is the remote only Infinispan (new style), and embedded caches are out of scope
* Out of scope: docs (as everything is still experimental)
* Test suite should run in KCB and in Keycloak to accept the new compound keys (offline flag + UUID key)

### Motivation

The motivation is to reduce the number of caches to configure and try to reduce the amount of memory (each cache has its overhead).
Use cases like removing realm or users will require fewer network calls due to the number of caches it needs to check.

### Details

Composite key, using tuple `<uuid, offline>`, will help identify if the session is online or offline.

### PR DESCRIPTION
**Pull request by pruivo**
Use a single cache for online and offline sessions to reduce the number of RPC calls.

Fixes #30932

### CODE REVIEW

**Review by ryanemerson — COMMENTED** [review]
A couple of very minor things, but LGTM.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/SessionKey.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.Objects;
+
+import org.infinispan.protostream.annotations.Proto;
+import org.infinispan.protostream.annotations.ProtoTypeId;
+import org.keycloak.marshalling.Marshalling;
+import org.keycloak.models.utils.KeycloakModelUtils;
+
+@ProtoTypeId(Marshalling.SESSION_KEY)
+@Proto
+public record SessionKey(String uuid, boolean offline) {
```
Comment (ryanemerson) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/SessionKey.java:29]: Does it make sense to just store this as a `UUID` given that it's wire format is just two longs?

File: model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java [hunk 2]
```diff
@@ -244,7 +244,10 @@ private RemoteCacheManager initRemote() {
         RemoteCacheManager remoteCacheManager = new RemoteCacheManager(builder.build());
 
         // establish connection to all caches
-        Arrays.stream(CLUSTERED_CACHE_NAMES).forEach(remoteCacheManager::getCache);
+        Arrays.stream(CLUSTERED_CACHE_NAMES)
+                .filter(name -> !OFFLINE_CLIENT_SESSION_CACHE_NAME.equals(name))
```
[the remark this answers is not in the document]
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java:248]: In production, It is only invoked once at boot time. It doesn't make sense to cache it, IMO.

[the remark this answers is not in the document]
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java:248]: I think I have a static method with the filter for the test suite. I can use it here too.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/SessionKey.java (hunk 1 above)
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/SessionKey.java:29]: To reduce the network traffic and storage space, yes. 

My concerns are around queries and indexes. It is unclear if I need to index this field or if `UUID` can be indexed. Also, the Keycloak code uses `String` everywhere. Example: 
https://github.com/keycloak/keycloak/blob/18a6c790113998bf039f46d6c71ffb668e961036/server-spi/src/main/java/org/keycloak/models/UserSessionModel.java#L38

File: model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java]: updated!

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/SessionKey.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    public SessionKey(String uuid, boolean offline) {
+        this.uuid = Objects.requireNonNull(uuid);
+        this.offline = offline;
+    }
+
+    public static SessionKey randomOnlineSessionKey() {
+        return randomSessionKey(false);
+    }
+
+    public static SessionKey randomOfflineSessionKey() {
```
[the remark this answers is not in the document]
Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/SessionKey.java]: removed!

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/SessionKey.java (hunk 1 above)
Comment (ryanemerson) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/SessionKey.java:29]: If we want indexing, we probably would have to add our own UUIDAdapter. I'm happy leaving this as a String.

Comment (pruivo) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/SessionKey.java:29]: ProtoStream ships the UUIDAdapter and I'm not sure which one the Infinispan Server will pick. I would rather stay with String for now :+1:

**Comment by sschu** [comment]
@pruivo I haven't looked at code but I would just like to make sure that after merging it is still possible to have the current Keycloak behaviour i.e. normal sessions only in cache and offline sessions persisted in DB and lazy-loaded into the cache?
Also, a downside I see is that it was formerly possible to get the number of normal and offline sessions by looking at the cache metrics - I guess this doesn't work anymore...

**Comment by ahus1** [comment]
> Also, a downside I see is that it was formerly possible to get the number of normal and offline sessions by looking at the cache metrics - I guess this doesn't work anymore...

@sschu/@pruivo - if you are using this "remote-cache" feature after this PR, you will not be able to distinguish in your metrics between online and offline sessions. 

About the change in behavior: With "remote-cache" enabled (even today in main), an offline session will no longer be stored in the database, but in the external Infinispan only _unless_ you enable "persistent-user-sessions".

**Comment by pruivo** [comment]
@ahus1 & @sschu & @mhajas is the number of offline/online sessions metric important enough to reject this PR?

**Comment by sschu** [comment]
The other thing that came to my mind is that you might want to configure cache settings differently for online and offline sessions also in the external Infinispan e.g. persistence.

**Comment by ahus1** [comment]
I agree with @sschu that persistence is possibly the thing that is most likely to be configured differently. I'm adding the label "hold" for now until we resolve this in this discussion.

Configuring different persistence for offline and online user sessions is something that IMHO is most likely the thing people might want to do. At the same time, it might be something due to the past as offline sessions were the only ones kept in a database with older versions of Infinispan until persistent sessions arrived. 

The original use case for offline and online sessions is when a token handed to an application is still valid. With an online session, the tokens are only valid while the user is logged in usually via their browsers - and this is where they then log out and they log out from all applications they have been using with that session. Offline sessions provide a token to an application that survives that logout and it might be handy for example for a mobile application that run independently of the daily login cycle. 

So if we aim to make persistence configuration with Infinispan simple, lightweight and "just works" for an external Infinispan, we'd be IMHO better off when we simplify things to a single cache. 

When it comes to getting information how many offline sessions there are, I am not sure if this is something you would need to do for monitoring continuously, of if it can be something where you query Infinispan now and then. Once we push the Potostream schema to Infinispan, you'll be able to run queries against the sessions in the Infinispan UI and I assume it would be possible via Infinispan's REST API as well. The advantage would be that it would be simpler for you to figure out as well which client created that offline sessions. Something we would need to consider: once we start to document that, it will be a public and supported way of interacting with Infinispan. 

It might be time to involve the @keycloak/core-iam  and @keycloak/core-clients team in this to get more input.

**Comment by ryanemerson** [comment]
> When it comes to getting information how many offline sessions there are, I am not sure if this is something you would need to do for monitoring continuously, of if it can be something where you query Infinispan now and then. Once we push the Potostream schema to Infinispan, you'll be able to run queries against the sessions in the Infinispan UI and I assume it would be possible via Infinispan's REST API as well.

+1 to providing a dedicated query capability for such information. We shouldn't rely on the cache sizes provided by metrics to provide this information as it's an expensive operation due to caches being distributed, and by default Infinispan only returns approximate values. See [ISPN-12607](https://issues.redhat.com/browse/ISPN-12607) and [ISPN-13421](https://issues.redhat.com/browse/ISPN-13421) for more context.

**Comment by pruivo** [comment]
I don't have any strong argument to keep this PR around. Dropping the offline caches would benefit the embedded cache the most by reducing the memory (long and short-live objects). 

For remote-only caches, only a couple of use cases benefit from the reduced number of caches: remove user and remove realm. Those are not so frequently used or are strong enough reasons to merge this change.

**Comment by sschu** [comment]
For the embedded caches, I would probably want to configure cache sizes or evictions differently since the usage patterns of normal and offline tokens are very different. So there I would definitely want to keep them separate. Long-term, there is also the isue that having several offline_client_session for one offline_user_session does not make sense. Two completely different mobile apps on the same device might get tied to the same offline_user_session and when one of the apps revokes this affects the other app's token. So maybe in the future there will be just a single offline_token cache anyways.

**Comment by mhajas** [comment]
@sschu Could you please elaborate more on "the usage patterns of normal and offline tokens are very different"? I never fully understood why we are distinguishing between those but maybe I am lacking some real-world example where it makes sense.

Why would users want to configure persistence only for offline sessions but not online sessions? Is it to have offline sessions that survive restarts but users are ok to lose online sessions? Is it because they live longer and therefore it is expected there is a higher amount of offline sessions and we want to save memory? Or is there something else? 

Understanding this may be an important thing also for the persistent user sessions feature, as what you are suggesting is to preserve the behaviour we have currently by default and goes the opposite direction we are currently implementing with persistent sessions.

**Comment by sschu** [comment]
@mhajas In our use cases, offline tokens are mostly used so people with mobile apps don't have to re-authenticate. They typically have infinite session lifetime, 180 days idle timeout (so there is plenty of time for them not using the app) and only small windows of increased activity (e.g. when traveling). Therefore we would like to push these sessions out of the caches quickly after their activity window has passed. 

Normal sessions are mostly the typical you start using you web applications in the morning and over night the session expires and its fine to just keep the session in the cache that long.

With persistent user sessions, you probably would not need offline sessions anymore for the first use case. Still, the usage pattern would differ depending on whether your app is something people use continuously/daily or only for special activities.

When offline tokens are really used to do something from the backend when the user is not present, I would expect a much more regular usage pattern since any actions might be time-triggered. But that is more of a speculation, we don't have these use cases, yet.

**Comment by ahus1** [comment]
AFAIK Infinispan uses a "least frequently used" algorithm to evict entries. Maybe @pruivo could comment on that. IMHO it has the potential to solve the situation @sschu describes without the need to manually configure things in different caches.

**Comment by sschu** [comment]
@ahus1 LFU is still an odd choice because new stuff is continuously evicting each other while old stuff is sitting comfortably in the cache. So I would have expected LRU as a default for just about everything. ;) But maybe I am getting something wrong here.

**Comment by ahus1** [comment]
@sschu - LFU is the default in Infinispan anyway. While I still wait for someone to explain it to me, I assume it is handling the case you describe.

**Comment by pruivo** [comment]
@sschu @ahus1 Caffeine only supports tiny LFU: https://github.com/ben-manes/caffeine/wiki/Efficiency

**Comment by sschu** [comment]
I had a quick glimpse and it looks like that it's much different from simple LFU which would just count total numbers. It seems to take recency into account like LRU which probably makes it a good fit.

**Comment by ahus1** [comment]
I clicked the link, found a wall of text and stayed with "I assume it is handling the case". :shrug:

**Comment by pruivo** [comment]
I'm dropping this PR. Having the possibility of configuring the offline sessions cache differently, is a good enough reason to drop this PR. 

Thanks for everyone's feedback!

**Pull request closed without merging**